### ISSUE TITLE
Another iterator suggestion.

### ISSUE THREAD

**Issue by jenstroeger** [more-itertools/more-itertools#468]
Iterators are very useful for comprehensions, and today I came across a problem that just happened to be solveable with [prepend()](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.prepend) but there is no function that returns an iterator for the more general case.

I have a bunch of values, some scalar and some iterable, and I need to iterate over the scalars and the first level of the iterables:
```python
def didum(*args):
    for value in args:
        if isinstance(value, (str, bytes,)):
            yield value
            continue
        if not value:
            yield value
            continue
        try:
            yield from value
        except TypeError:
            yield value
```
This one’s useful because it doesn’t create temporary data structures (like prepend() does). You can then call this function to iterate over its argument values:
```
>>> list(didum(1, 2, "foo"))
[1, 2, 'foo']
>>> list(didum(b"bar", [1, 2, 3], 4))
[b'bar', 1, 2, 3, 4]
>>> list(didum([], "", 0, None))
[[], '', 0, None]
>>> list(didum())
[]
```
The existing prepend() can then be implemented as:
```
def prepend(value, iterator):
    return didum(value, iterator)  # Don’t create a [value] list.
```
I guess it’s a combination of generalizing prepend() and [flatten()](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.flatten). I have no name for it though 😉

If this is interesting for you, I’d be happy to submit a PR. If this exists already, I guess I’ve missed it.

**Comment by Ricocotam** [comment]
I was actually going to post a similar iterator that I called `flatten_iflist_iterate`

```python
def flatten_iflist_iterate(input_list):
    for item in input_list:
        if type(item) == list:
            for i in flatten_iflist_iterate(item):
                yield i
        else:
            yield item
```
I think your code is actually better but we need a recursion to handle nested lists : 

```python
>>> a = [1, 2, [3, 4, [5, [6], 7]], 8, 9]
>>> list(flatten_iflist_iterate(a))
[1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> list(didium(a))
[1, 2, 3, 4, [5, [6], 7], 8, 9]
```

**Comment by jenstroeger** [comment]
@Ricocotam, I actually didn’t want to recurse into nested iterators and only skim the first level.
```
>>> a = [1, 2, [3, 4, [5, [6], 7]], 8, 9]
>>> list(didum(a))
[1, 2, 3, 4, [5, [6], 7], 8, 9]
```
Yes that is intended because you're passing a single argument that’s the only value returned by the iterator.

Adding flattening nested iterators will add more complexity that I didn’t need, but I guess you could fall back onto the existing [collapse()](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.collapse) function:
```python
def didum(*args, collapse=False):
    for value in args:
        if isinstance(value, (str, bytes,)):
            yield value
            continue
        if not value:
            yield value
            continue
        try:
            if collapse:
                yield from more_itertools.collapse(value)
            else:
                yield from value
        except TypeError:
            yield value
```
Which then gives you:
```
>>> list(didum(a))
[1, 2, [3, 4, [5, [6], 7]], 8, 9]
>>> list(didum(a, collapse=True))
[1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> list(didum(a, a, a, collapse=True))
[1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

**Comment by bbayles** [comment]
The name that comes to mind: `value_chain`. What do you think?

**Comment by jenstroeger** [comment]
@bbayles, oh I like that, yes! Would you like a PR for this iterator?

**Comment by bbayles** [comment]
Yes, I'm interested in the non-recursive version. Thanks!

**Comment by Ricocotam** [comment]
I think both should exists for different purpose

**Comment by jenstroeger** [comment]
@Ricocotam, does my second attempt [above](https://github.com/more-itertools/more-itertools/issues/468#issuecomment-706509633) address your needs? I am still unclear: isn’t [collapse()](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.collapse) what you’re looking for?

**Comment by Ricocotam** [comment]
Well yes it does fit well what I was looking for. Thanks for the suggestion 👍

**Comment by bbayles** [comment]
> And @bbayles, is this one a recipe or iterator?

Original contributions go in `more.py`.

**Comment by bbayles** [comment]
Merged in PR #75. Many thanks!
